vendor/vendor-android/src/main/kotlin/com/malinskiy/marathon/android/executor/listeners/TestRunResultsListener.kt

```diff
@@ -50,56 +54,85 @@ class TestRunResultsListener(private val testBatch: TestBatch,
+                        logger.warn { "infraFailure = ${result.test.toTestName()}, ${device.serialNumber}" }
+                        infraFailures.add(result)
+                    } else {
+                        failed.add(result)
+                    }
                 }
-                .values
-                .createUncompletedTestResults(runResult, device)
 
-        if (uncompleted.isNotEmpty()) {
-            uncompleted.forEach {
-                logger.warn { "uncompleted = ${it.test.toTestName()}, ${device.serialNumber}" }
+                DdmLibTestStatus.INCOMPLETE, null -> {
+                    logger.warn { "uncompleted = ${result.test.toTestName()}, ${device.serialNumber}" }
+                    incomplete.add(result)
+                }
             }
         }
 
-        deferred.complete(TestBatchResults(device, finished, failed, uncompleted))
+        val noStatus = tests
+            .filterNot { expectedTest ->
+                results.containsKey(expectedTest.key)
+            }
+            .values
+            .createUncompletedTestResults(runResult, device)
+
+        noStatus.forEach {
+            logger.warn { "noStatus = ${it.test.toTestName()}, ${device.serialNumber}" }
+        }
+
+        logger.debug(
+            "Batch test results: " +
+                    "passed=${passed.size}; " +
+                    "failed=${failed.size}; " +
+                    "infraFailures=${infraFailures.size}; " +
+                    "uncompleted=${incomplete.size}; " +
+                    "noStatus=${noStatus.size}"
+        )
+
+        deferred.complete(
+            TestBatchResults(
+                device, passed, failed,
+                uncompleted = infraFailures + incomplete + noStatus
+            )
+        )
     }
 
-    private fun Collection<Test>.createUncompletedTestResults(testRunResult: com.android.ddmlib.testrunner.TestRunResult,
-                                                              device: Device): Collection<TestResult> {
+    private fun Collection<Test>.createUncompletedTestResults(
+        testRunResult: com.android.ddmlib.testrunner.TestRunResult,
+        device: Device
+    ): Collection<TestResult> {
 
         val lastCompletedTestEndTime = testRunResult
-                .testResults
-                .values
-                .maxBy { it.endTime }
-                ?.endTime
-                ?: timer.currentTimeMillis()
+            .testResults
+            .values
+            .maxBy { it.endTime }
+            ?.endTime
+            ?: timer.currentTimeMillis()
 
         return map {
             TestResult(
-                    it,
-                    device.toDeviceInfo(),
-                    TestStatus.FAILURE,
-                    lastCompletedTestEndTime,
-                    lastCompletedTestEndTime,
-                    testRunResult.runFailureMessage
+                it,
+                device.toDeviceInfo(),
+                TestStatus.INCOMPLETE,
```

Comment:
Last time, I failed to fully comprehend what 'uncompleted test results' meant. Still do. But just to be sure, are you confident changing failure here to incomplete wouldn't break how the results are handled in terms of re-queueing and de-queueing them?

Reply:
1. Uncompleted means that test will be retried - and that retry won't be counted towards the limit of 5 retries. But it will count towards the global limit of such retries.
2. I am not 100% - that's why I am going to check new Marathon version extensively. But I think that actually it will be better in terms of re-queueing and de-queueing. There is no magic in it - all decisions are made based not on status, but on which list contains test result (finished, failed or uncompleted).